pkg/compactor/block_upload.go

```diff
@@ -528,6 +542,32 @@ func (c *MultitenantCompactor) validateBlock(ctx context.Context, blockID ulid.U
 	return nil
 }
 
+func (c *MultitenantCompactor) validateMaximumBlockSize(logger log.Logger, files []metadata.File, userID string) error {
+	maxBlockSizeBytes := c.cfgProvider.CompactorBlockUploadMaxBlockSizeBytes(userID)
+	if maxBlockSizeBytes <= 0 {
+		return nil
+	}
+
+	blockSizeBytes := int64(0)
+	for _, f := range files {
+		if f.SizeBytes < 0 {
+			return errors.New("invalid negative file size in block metadata")
+		}
+		blockSizeBytes += f.SizeBytes
+		if blockSizeBytes < 0 {
+			// overflow
+			break
+		}
+	}
+
+	if blockSizeBytes > maxBlockSizeBytes || blockSizeBytes < 0 {
+		err := fmt.Errorf(maxBlockUploadSizeBytesFormat, maxBlockSizeBytes)
+		level.Error(logger).Log("err", err, "size", blockSizeBytes)
```

Comment:
wouldn't this error be also logged here? https://github.com/grafana/mimir/blob/bf1be656332427a91748c13ac98d7f7e0d91097b/pkg/compactor/block_upload.go#L313

Reply:
> WDYT about making this option advanced? 

All for it, thanks I hadn't thought about the category when originally writing this.

> wouldn't this error be also logged here? 

I'll look into this. The only reason I added this log line was to give visibility into the observed block size, but I didn't want to communicate that back to the user.

Reply:
> I'll look into this. The only reason I added this log line was to give visibility into the observed block size, but I didn't want to communicate that back to the user.

i wouldn't mind having a more verbose error for the user. I was more concerned about noisy logging. But not feeling strongly about this, so however you decide

Reply:
Agree that it seems unnecessary to hide computed block size from user.

Reply:
I generally defer to providing less information in errors to reduce adding surface area to an API. I thought logging on the server side was more useful because the rejections could be analyzed all together, but the information provided to the user is less relevant and possibly even confusing (the metadata size isn't counted, an overflow would return a negative number and may not be representative of the entire sum).

Reply:
> I generally defer to providing less information in errors to reduce adding surface area to an API. I thought logging on the server side was more useful because the rejections could be analyzed all together, but the information provided to the user is less relevant and possibly even confusing (the metadata size isn't counted, an overflow would return a negative number and may not be representative of the entire sum).

I don't mind if we report overflown value to the user (realistically that will never happen for "normal" blocks), and we can avoid confusion by better wording saying that it doesn't include block metadata.

The point is that we already log errors returned to user, so why bother logging it once more with extra details, if this detail is something user can figure out themselves and is not sensitive in any way.

(But this is a nit, we can keep it as is)

Reply:
I was thinking about this some more. Rather than returning the size of the rejection maybe a histogram of block sizes proposed in start upload would be generally more useful. If it included rejected sizes as well it could provide insight on what the limit could be moved to without the burden of interpreting log lines/errors. Overflow might have to be excluded to avoid negatives though, but that seems fine.

Reply:
> Rather than returning the size of the rejection maybe a histogram of block sizes proposed in start upload would be generally more useful.

Histogram will only give us aggregated values, which is fine, but we can also log all block sizes to have specific values.